src/core/layertree/qgslayertreelayer.cpp

```diff
@@ -173,9 +173,14 @@ void QgsLayerTreeLayer::layerWillBeDeleted()
 
 }
 
-
 void QgsLayerTreeLayer::layerNameChanged()
 {
   Q_ASSERT( mRef );
   emit nameChanged( this, mRef->name() );
 }
+
+void QgsLayerTreeLayer::setExpression( const QString &expression )
+{
+  mExpression = expression;
+}
+
```

Comment:
Adding a function to expose the expression member, this member stores the expression associated with the layer to ensure that every layer can have their own expression without conflict in case of duplicates.